### PR TITLE
PMM-4131 Port features from dcu/mongodb_exporter.

### PR DESCRIPTION
**Pull request by DexterHD**
- [x] tcmalloc.go
- [x] conn_pool_stats.go
- [x] replset_conf.go
- [x] ~~collection_wiredtiger.go~~

### CODE REVIEW

File: collector/common/conn_pool_stats.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	numAScopedConnections.Set(stats.ASScopedConnections)
+	numAScopedConnections.Collect(ch)
+
+	totalInUse.Set(stats.TotalInUse)
+	totalInUse.Collect(ch)
+
+	totalAvailable.Set(stats.TotalAvailable)
+	totalAvailable.Collect(ch)
+
+	ch <- prometheus.MustNewConstMetric(totalCreatedDesc, prometheus.CounterValue, stats.TotalCreated)
```
Comment (AlekSi) [inline collector/common/conn_pool_stats.go]: Why this one metric is not like the others?
If it is because it is a counter, then constant counters can be created using the same method. As a bonus point that will not introduce new mutable package-level variables that you wanted to remove.

The same in other places.

File: collector/common/conn_pool_stats.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	totalAvailable.Describe(ch)
+	ch <- totalCreatedDesc
+}
+
+// GetConnPoolStats returns the server connPoolStats info.
+func GetConnPoolStats(client *mongo.Client) *ConnPoolStats {
+	result := &ConnPoolStats{}
+	err := client.Database("admin").RunCommand(context.TODO(), bson.D{{"connPoolStats", 1}, {"recordStats", 0}}).Decode(result)
+	if err != nil {
+		log.Error("Failed to get server status.")
```
Comment (AlekSi) [inline collector/common/conn_pool_stats.go]: Let's log `err`.

The same in other places.

File: collector/common/tcmalloc.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	tcmallocCacheBytes.WithLabelValues("thread_cache", "max_total").Set(m.Details.MaxTotalThreadCacheBytes)
+	tcmallocCacheBytes.WithLabelValues("thread_cache", "current_total").Set(m.Details.CurrentTotalThreadCacheBytes)
+	tcmallocCacheBytes.WithLabelValues("central_cache", "free").Set(m.Details.CentralCacheFreeBytes)
+	tcmallocCacheBytes.WithLabelValues("transfer_cache", "free").Set(m.Details.TransferCacheFreeBytes)
+	tcmallocCacheBytes.WithLabelValues("thread_cache", "free").Set(m.Details.ThreadCacheFreeBytes)
+	tcmallocCacheBytes.Collect(ch)
+
+	ch <- prometheus.MustNewConstMetric(tcmallocAggressiveDecommitDesc, prometheus.CounterValue, m.Details.AggressiveMemoryDecommit)
+	ch <- prometheus.MustNewConstMetric(tcmallocFreeBytesDesc, prometheus.CounterValue, m.Details.TotalFreeBytes)
+
```
Comment (AlekSi) [inline collector/common/tcmalloc.go]: Empty line.

File: collector/mongod/replset_conf.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
+package mongod
+
+import (
+	"context"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+
```
Comment (AlekSi) [inline collector/mongod/replset_conf.go]: Merge those groups.

File: collector/common/conn_pool_stats_test.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/percona/mongodb_exporter/testutils"
+)
+
+func TestGetConnPoolStatsDecodesFine(t *testing.T) {
+	// setup
+	t.Parallel()
```
Comment (AlekSi) [inline collector/common/conn_pool_stats_test.go]: That should be moved to subtests.

File: collector/common/conn_pool_stats_test.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+	defer cancel()
+
+	t.Run("mongod", func(t *testing.T) {
+		// setup
+		defaultClient := testutils.MustGetConnectedMongodClient(ctx, t)
+		defer defaultClient.Disconnect(ctx)
+		// run
+		statusDefault := GetConnPoolStats(defaultClient)
+		// test
+		assert.NotNil(t, statusDefault)
```
Comment (AlekSi) [inline collector/common/conn_pool_stats_test.go]: Can we check at least _some_ values? This test (and others) will happily pass if an empty structure is returned.

File: collector/mongod/server_status.go [hunk 7]
```diff
@@ -22,12 +22,12 @@ import (
...
 	"go.mongodb.org/mongo-driver/mongo"
 
-	collector_common "github.com/percona/mongodb_exporter/collector/common"
+	commoncollector "github.com/percona/mongodb_exporter/collector/common"
 )
 
 // ServerStatus keeps the data returned by the serverStatus() method.
 type ServerStatus struct {
-	collector_common.ServerStatus `bson:",inline"`
+	commoncollector.ServerStatus `bson:",inline"`
```
Comment (AlekSi) [inline collector/mongod/server_status.go]: Why not `common` without named import?

File: mongodb_exporter.go [hunk 8]
```diff
@@ -42,10 +42,11 @@ var (
...
 
-	collectDatabaseF   = kingpin.Flag("collect.database", "Enable collection of Database metrics").Bool()
-	collectCollectionF = kingpin.Flag("collect.collection", "Enable collection of Collection metrics").Bool()
-	collectTopF        = kingpin.Flag("collect.topmetrics", "Enable collection of table top metrics").Bool()
-	collectIndexUsageF = kingpin.Flag("collect.indexusage", "Enable collection of per index usage stats").Bool()
+	collectDatabaseF            = kingpin.Flag("collect.database", "Enable collection of Database metrics").Bool()
+	collectCollectionF          = kingpin.Flag("collect.collection", "Enable collection of Collection metrics").Bool()
+	collectTopF                 = kingpin.Flag("collect.topmetrics", "Enable collection of table top metrics").Bool()
+	collectIndexUsageF          = kingpin.Flag("collect.indexusage", "Enable collection of per index usage stats").Bool()
+	mongodbCollectConnPoolStats = kingpin.Flag("collect.connpoolstats", "Collect MongoDB connpoolstats").Bool()
```
Comment (AlekSi) [inline mongodb_exporter.go]: Add `F` suffix as in other flags. Also make help text consistent with other `collect.XXX` flags.

File: collector/mongod/server_status.go (hunk 7 above)
Comment (DexterHD) [reply collector/mongod/server_status.go]: To add a specific sense. I decide that the word `common` by itself doesn't look good in this context.

File: collector/common/conn_pool_stats_test.go (hunk 6 above)
Comment (DexterHD) [reply collector/common/conn_pool_stats_test.go]: Probably yes, but for this: 

1. We should force MongoDB to increase that numbers, as at least in the local environment all numbers returned by this command have `0` value.
2. We should change the test name.

File: collector/common/conn_pool_stats.go (hunk 1 above)
Comment (DexterHD) [reply collector/common/conn_pool_stats.go]: > Why this one metric is not like the others?

Yes, it's because it is a counter.

> If it is because it is a counter, then constant counters can be created using the same method.

Sorry but I don't understand what exactly did you mean. Can you rephrase it? Should I change all other metrics by constant metrics or something else?

Comment (AlekSi) [reply collector/common/conn_pool_stats.go]:
```go
ch <- prometheus.MustNewConstMetric(syncClientConnectionsDesc, prometheus.GaugeValue, stats.SyncClientConnections)
```

Basically, https://jira.percona.com/browse/PMM-4167, but only for the new code in this PR.

File: collector/common/conn_pool_stats_test.go (hunk 6 above)
Comment (AlekSi) [reply collector/common/conn_pool_stats_test.go]: Let's do that.

Comment (DexterHD) [reply collector/common/conn_pool_stats_test.go]: > We should force MongoDB to increase that numbers,

I've read the docs and to do that we should configure MongoDB in `replSet` or `cluster` mode, and connect together 2 or more servers. I think it some kind of overengineering for our testing environment at least in this case.

> This test (and others) will happily pass if an empty structure is returned.

To handle this case I've found than MongoDB [has a special field](https://docs.mongodb.com/manual/tutorial/use-database-commands/#command-responses) called `ok` for all responses that were succeeded. I've added this field in our tests. Now those tests wouldn't pass if an empty struct returned.

File: collector/mongos/database_status.go [hunk 9]
```diff
@@ -109,15 +109,15 @@ func GetDatabaseStatList(client *mongo.Client) *DatabaseStatList {
...
+		log.Error(err)
 		return nil
 	}
 	for _, db := range dbNames {
 		dbStatus := DatabaseStatus{}
 		r := client.Database(db).RunCommand(context.TODO(), bson.D{{"dbStats", 1}, {"scale", 1}})
 		err := r.Decode(&dbStatus)
 		if err != nil {
-			log.Error("Failed to get database status.")
+			log.Error(err)
```
Comment (AlekSi) [inline collector/mongos/database_status.go]: That will log something like "context deadline exceeded" without any context (pun intended). I would say we should use both message and error:
```go
log.Errorf("Failed to get database status: %s.", err)
```

File: collector/mongos/database_status.go [hunk 10]
```diff
@@ -109,15 +109,15 @@ func GetDatabaseStatList(client *mongo.Client) *DatabaseStatList {
 	dbStatList := &DatabaseStatList{}
 	dbNames, err := client.ListDatabaseNames(context.TODO(), bson.M{})
 	if err != nil {
-		log.Error("Failed to get database names")
+		log.Error(err)
```
Comment (AlekSi) [inline collector/mongos/database_status.go]: One more place.

File: collector/mongos/sharding_changelog.go [hunk 11]
```diff
@@ -104,7 +104,7 @@ func GetShardingChangelogStatus(client *mongo.Client) *ShardingChangelogStats {
 
 	c, err := coll.Aggregate(context.TODO(), []bson.M{{"$match": match}, {"$group": group}})
 	if err != nil {
-		log.Error("Failed to execute find query on 'config.changelog'!")
+		log.Error(err)
```
Comment (AlekSi) [inline collector/mongos/sharding_changelog.go]: And there.

**Comment by AlekSi** [comment]
@BupycHuk PTAL

File: collector/common/conn_pool_stats_test.go (hunk 6 above)
Comment (BupycHuk) [reply collector/common/conn_pool_stats_test.go]: But why they all have 0 value?

Comment (DexterHD) [reply collector/common/conn_pool_stats_test.go]: Because you need to have more than 1 server. Those metrics related to `replSet` or `cluster` mode, so to have connection pool you should have configured cluster first. 😄

File: collector/mongos/server_status_test.go [hunk 12]
```diff
@@ -77,3 +81,17 @@ func loadServerStatusFromBson(data []byte, status *ServerStatus) {
 		panic(err)
 	}
 }
+
+func TestGetServerStatusDecodesFine(t *testing.T) {
+	// setup
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```
Comment (BupycHuk) [inline collector/mongos/server_status_test.go]: Why do we need this timeout?

Comment (DexterHD) [reply collector/mongos/server_status_test.go]: 1) It's [standard approach](https://github.com/mongodb/mongo-go-driver#usage) to working with the new mongodb_driver.
2) To force disconnecting from MongoDB if something goes wrong. Without it, there is a potential risk that tests can become stuck.
